R/utils.R

```diff
@@ -504,3 +504,175 @@ download_with_retry <- function(...) {
+check_nonmem_finished <- function(mod) {
+  UseMethod("check_nonmem_finished")
+}
+
+
+#' Check if NONMEM is finished (character)
+#' @param mod character string of NONMEM model
+#'
+#' @importFrom checkmate assert_string
+#' @importFrom readr read_lines
+#'
+#' @describeIn check_nonmem_finished takes a file path (character string) of `NONMEM` `.lst` file.
+#' @export
+check_nonmem_finished.character <- function(mod) {
+  assert_string(mod)
+
+  # look for model to be finished and then test output
+  model_finished <- try({suppressWarnings({
+    read_lines(mod) %>%
+      str_detect("Stop Time") %>%
+      any()
+  })})
+  return(isTRUE(model_finished))
+}
+
+#' Check if NONMEM is finished (bbi model)
+#' @param mod bbi model object
+#'
+#' @describeIn check_nonmem_finished takes a `bbi_nonmem_model` object.
+#' @export
+check_nonmem_finished.bbi_nonmem_model <- function(mod) {
+  # check that output dir was created (should happen as soon as model is submitted)
+  if (!fs::dir_exists(get_output_dir(mod, .check_exists = FALSE))) {
+    return(TRUE) # if missing then this failed right away, likely for some bbi reason
+  }
+
+  # pass .lst path to character dispatch
+  build_path_from_model(mod, ".lst") %>%
+    check_nonmem_finished()
+}
+
+#' Wait for NONMEM models to finish
+#' @param mod a `bbi_nonmem_model` object, or list of `bbi_nonmem_model` objects.
+#' @param time_limit integer for maximum number of seconds in total to wait before continuing (exiting function)
+#' @param interval integer for number of seconds to wait between each check
+#'
+#'
+#' @description Calling `wait_for_nonmem()` will freeze the user's console until the model(s) have finished running.
+#'
+#' @export
+wait_for_nonmem <- function(mod, time_limit = 200, interval = 5) {
+  UseMethod("wait_for_nonmem")
+}
+
+
+#' Wait for NONMEM models to finish  (bbi model)
+#' @param mod a `bbi_nonmem_model` object, or list of `bbi_nonmem_model` objects.
+#' @param time_limit integer for maximum number of seconds in total to wait before continuing (exiting function)
+#' @param interval integer for number of seconds to wait between each check
+#'
+#' @describeIn wait_for_nonmem takes a `bbi_nonmem_model` object.
+#' @export
+wait_for_nonmem.bbi_nonmem_model <- function(mod, time_limit = 200, interval = 5) {
+  wait_for_nonmem(list(mod), time_limit = time_limit, interval = interval)
+}
+
+
+#' Wait for NONMEM models to finish  (list of bbi models)
+#' @param mod a `bbi_nonmem_model` object, or list of `bbi_nonmem_model` objects.
+#' @param time_limit integer for maximum number of seconds in total to wait before continuing (exiting function)
+#' @param interval integer for number of seconds to wait between each check
+#'
+#' @importFrom purrr map_lgl
+#' @importFrom checkmate assert_list
+#'
+#'
+#' @describeIn wait_for_nonmem takes a `list` of `bbi_nonmem_model` objects.
+#' @export
+wait_for_nonmem.list <- function(mod, time_limit = 200, interval = 5) {
+
+  assert_list(mod)
```

Comment:
Probably want to throw in a [`check_model_object_list(mod)`](https://github.com/metrumresearchgroup/bbr/blob/main/R/utils.R#L385-L395) (as used in [`submit_models()`](https://github.com/metrumresearchgroup/bbr/blob/develop/R/submit-models.R#L44) for example) just to be safe.